### ISSUE TITLE
[BUG] generic "No OpenSearchException found" error if http.detailed_errors is disabled

### ISSUE THREAD

**Issue by rursprung** [opensearch-project/OpenSearch#4545]
**Describe the bug**
if `http.detailed_errors.enabled: false` is set then the error message in a REST API call will only report back an error message if the exception contains an `OpenSearchException`:
https://github.com/opensearch-project/OpenSearch/blob/465771631d2e3a5001d97d0c7e845e6f8ea6bd15/server/src/main/java/org/opensearch/OpenSearchException.java#L595-L608

(note: confusingly enough this method is a static member of `OpenSearchException` but handles any `Exception`, IMHO it might be better located e.g. in [ExceptionsHelper](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/ExceptionsHelper.java)?)

there are however (many?) cases where this isn't the case, e.g. when trying to update a setting through the settings API which doesn't support it (fun-fact: i noticed it when trying to update the `http.detailed_errors.enabled` setting). this returns an `IllegalArgumentException` instead.

**To Reproduce**
1. start a cluster with `http.detailed_errors.enabled: false`
2. call `PUT /_cluster/settings` with a setting which doesn't support it, e.g.:
   ```json
   {
       "persistent": {
           "http.detailed_errors.enabled": true
       }
   }
   ```

with `http.detailed_errors.enabled: false` (as in the reproduction case) it gives the following - useless - response:
```json
{
    "error": "No OpenSearchException found",
    "status": 400
}
```

with the default settings (or explicitly `http.detailed_errors.enabled: true`) it gives the following response:
```json
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_argument_exception",
                "reason": "persistent setting [http.detailed_errors.enabled], not dynamically updateable"
            }
        ],
        "type": "illegal_argument_exception",
        "reason": "persistent setting [http.detailed_errors.enabled], not dynamically updateable"
    },
    "status": 400
}
```

**Expected behavior**
the error message gives a hint at the cause as here the cause is clearly on the side of the caller. without the message the caller has no way of knowing what to do to fix the issue.

otherwise the message should at least be in the server log, but this isn't the case either (with default log settings). that way no information leakage would happen but it'd still be possible to understand the cause of a failure (for people with access to the system which should be the only ones which can do config updates via APIs).

**Plugins**
(irrelevant, this is about core functionality)

**Screenshots**
n/a

**Host/Environment (please complete the following information):**
OpenSearch 2.2.0 running in a linux container

**Additional context**
see the following related tickets:
* https://github.com/opensearch-project/OpenSearch/issues/4255
* https://github.com/opensearch-project/documentation-website/issues/1235

**Comment by xuezhou25** [comment]
Update my findings: When `http.detailed_errors.enabled` is set to false and the exception type is from JAVA exception, it will return general message "No OpenSearchException found".(Example from previous reproduction)

Providing another example with `http.detailed_errors.enabled: false` and exception is from OpenSearchException, it will response with `type` and `reason` of root cause:

```
$ curl -H 'Content-Type: application/json' -XPUT localhost:9200/_cluster/settings -d '{ "persistent" : { "cluster.blocks.read_only": true } }'

$ curl -XPUT "localhost:9200/new-index?pretty"
{
  "error" : "ClusterBlockException[index [new-index] blocked by: [FORBIDDEN/6/cluster read-only (api)];]",
  "status" : 403
}
```
If `http.detailed_errors.enabled` was set to `true` or with default setting
```
$ curl -XPUT "localhost:9200/new-index?pretty"

{
  "error" : {
    "root_cause" : [
      {
        "type" : "cluster_block_exception",
        "reason" : "index [new-index] blocked by: [FORBIDDEN/6/cluster read-only (api)];"
      }
    ],
    "type" : "cluster_block_exception",
    "reason" : "index [new-index] blocked by: [FORBIDDEN/6/cluster read-only (api)];"
  },
  "status" : 403
}
```

**Comment by andrross** [comment]
@reta @nknize Sorry for spamming folks here but looking for some guidance on a small but potentially impactful API change.

The [current behavior when `http.detailed_errors.enabled: false`](https://github.com/opensearch-project/OpenSearch/blob/465771631d2e3a5001d97d0c7e845e6f8ea6bd15/server/src/main/java/org/opensearch/OpenSearchException.java#L595-L608) is to only send exception type name and message in the API response if the exception type is OpenSearchException; otherwise the text "No OpenSearchException found" is sent. I can only guess that this is intentional. However, as @rursprung notes this can be quite painful as a user if the underlying cause is something like an IllegalArgumentException because you get no useful details beyond the HTTP error code. Changing the behavior to send the type name and message regardless of exception type feels fairly innocuous to me and would provide more useful details to the user, but I'm not sure about information leakage. I see a few options here (as suggested by @rursprung above):

1. Change the behavior to send the type name and message regardless of exception type
2. Add logging in the case where no details are sent (I don't love this option...feels a bit like the log-and-throw anti-pattern while also potentially spamming the log with 400-type user errors)
3. Play whack-a-mole at these cases throughout the code base where non-OpenSearchExceptions are thrown and change them to throw OpenSearchExceptions with helpful error messages
4. Do nothing; user gets an HTTP error code and the text "No OpenSearchException found"

Does anyone here have more historical context around this behavior? Any and all opinions are welcome.

**Comment by andrross** [comment]
Thanks @reta !

Looking at this a bit more, what do you think about mirroring the logic in [ExceptionHelpers.status()](https://github.com/opensearch-project/OpenSearch/blob/14704b30cafadcfdcd7cd81961ee07452804b0c3/server/src/main/java/org/opensearch/ExceptionsHelper.java#L87-L990) with something like the following?

```
public final class ExceptionsHelper {

    ...

    public static String summaryMessage(Throwable t) {
        if (t != null) {
            if (t instanceof OpenSearchException) {
                return t.getClass().getSimpleName() + "[" + t.getMessage() + "]";
            } else if (t instanceof IllegalArgumentException) {
                return "Invalid argument";
            } else if (t instanceof JsonParseException) {
                return "Failed to parse JSON";
            } else if (t instanceof OpenSearchRejectedExecutionException) {
                return "Too many requests";
            }
        }
        return "Internal failure";
    }
}
```

The logic in OpenSearchExceptions would become:

```
// Render the exception with a simple message
if (detailed == false) {
    Throwable t = e;
    for (int counter = 0; counter < 10 && t != null; counter++) {
        if (t instanceof OpenSearchException) {
            break;
        }
        t = t.getCause();
    }
    builder.field(ERROR, ExceptionHelpers.simpleMessage(t));
    return;
}
```

We'd mirror the logic around the known 400 errors with blessed messages. We'd also avoid exposing embarrassing details like a stray NullPointerException :)

**Comment by reta** [comment]
LGTM @andrross, I think it is reasonable alternative

**Comment by rursprung** [comment]
thanks for the discussion! the suggestion from @andrross would already be an improvement, but e.g. the settings API would then still not report anything useful (i.e. for a caller getting an error it'd still be unclear what he did wrong / what he'd have to change in order to make his call work). the problem is that `IllegalArgumentException` is a generic java exception - the code cannot know whether this was thrown in response to a user error (= should be reported to the user) or due to an internal system error / coding error leading to a wrong call to an API which then returns this error (= should maybe be kept back)

**Comment by andrross** [comment]
@rursprung I agree with your points. I think we should go ahead with the proposal above that will result in no longer responding with the confusing/unhelpful "No OpenSearchException found" text.

Separately, I think we should update the [AbstractScopedSettings class](https://github.com/opensearch-project/OpenSearch/blob/3aef125d0dbf2f29d9bf00b84cb52892a27a5328/server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java#L856) to throw OpenSearchExceptions instead of IllegalArgumentException so that the helpful error messages can be reported in all cases.

This keeps a fairly conservative approach of not reporting arbitrary exception details when `http.detailed_errors.enabled: false` while fixing the specific case of sending back helpful details when making invalid setting changes. Thoughts?

**Comment by rursprung** [comment]
i think that's a sensible trade-off, but it's then probably something which should be kept in mind when adding any new exceptions anywhere in the system: you always have to consider whether it should be an `OpenSearchException` or if it can be something else. and it might mean patching quite a few APIs over time if people encounter errors there and need a better error message.

**Comment by rursprung** [comment]
@xuezhou25 i saw that you implemented part of what was discussed here in #4669 (thanks for that!) and closed this issue here from there.
what about the second part mentioned by @andrross in https://github.com/opensearch-project/OpenSearch/issues/4545#issuecomment-1263908329:

> Separately, I think we should update the [AbstractScopedSettings class](https://github.com/opensearch-project/OpenSearch/blob/3aef125d0dbf2f29d9bf00b84cb52892a27a5328/server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java#L856) to throw OpenSearchExceptions instead of IllegalArgumentException so that the helpful error messages can be reported in all cases.

will you address this in a separate PR? should i raise a new ticket for this? note that it doesn't just affect the line linked in the comment, i count 17 `throw new IllegalArgumentException(...)` occurrences in `AbstractScopedSettings`.
what is the recommended way to go forward here? is it to now do `throw new OpenSearchException(new IllegalArgumentException(...))`? add a try/catch block around the method implementation, catch `IllegalArgumentException` and re-throw it wrapped in an `OpenSearchException`? or just throw an `OpenSearchException` instead of an `IllegalArgumentException`?

**Comment by xuezhou25** [comment]
@rursprung I created a new issue to throw new OpenSearchExceptions instead of IllegalArgumentException by updating in the AbstractScopedSettings class. 
Thanks for providing several solutions. I don't think do `throw new OpenSearchException(new IllegalArgumentException(...))` can meet the requirement because in the logic here 
https://github.com/opensearch-project/OpenSearch/blob/77cff55bd431c0e2f8fdea5a40c83f8e8bddeca1/server/src/main/java/org/opensearch/OpenSearchException.java#L593-L598
will find the root cause of exception and return "Invalid argument". So I am thinking of the way wrapping `IllegalArgumentException` to an `OpenSearchException`.
Keeping this issue closed.

**Comment by dblock** [comment]
https://github.com/opensearch-project/OpenSearch/issues/4745
